File: ios/banner/AdropPreloadedBannerStore.swift
import Foundation
import AdropAds

/// Registry of banners delivered by the batch `AdropBanner.loads()` path,
/// keyed by the JS-minted requestId. Shared between the AdropBanner module
/// (which fills and destroys entries) and `AdropPreloadedBannerView` (which
/// looks entries up on mount). Main-thread only (module methodQueue is main,
/// view props are set on main).
class AdropPreloadedBannerStore {
    static let instance = AdropPreloadedBannerStore()

    var banners: [String: AdropBanner] = [:]

    func requestIdFor(_ banner: AdropBanner) -> String {
        for entry in banners where entry.value === banner {
            return entry.key
        }
        return ""
    }
}

File: android/src/main/java/io/adrop/banner/AdropPreloadedBannerStore.kt
package io.adrop.banner

import io.adrop.ads.banner.AdropBanner
import java.util.concurrent.ConcurrentHashMap

/**
 * Registry of banners delivered by the batch `AdropBanner.loads()` path,
 * keyed by the JS-minted requestId. Shared between [io.adrop.AdropBannerModule]
 * (which fills and destroys entries) and [AdropPreloadedBannerViewManager]
 * (which looks entries up on mount).
 *
 * ConcurrentHashMap + main-thread mutation follows the module thread-safety
 * decision (react-native/docs/decisions/native-module-thread-safety.md).
 */
object AdropPreloadedBannerStore {
    val banners = ConcurrentHashMap<String, AdropBanner>()

    fun requestIdFor(banner: AdropBanner): String {
        banners.entries.find { it.value === banner }?.let { return it.key }
        return ""
    }
}
